hgshah/cmput404-project
# Complete follow endpoints with remote support!

Note: based on https://github.com/hgshah/cmput404-project/pull/99

# Overview

We can now do the full follow flow, also remotely!

## Changes

- Implement the follow flow from making a follow request, accepting it, getting followers, and even removing a follower or deleting a follow request
- Automated testing that we can run over and over to make sure our mirrored calls work all the time
- Check out endpoints.txt

## Testing
Just use automated testing!

The tests below will pretend to be a user in a browser, following each other.

To run:
1. Run this server in 8000 and 8080
2. From manage.py's directory:
    python manage.py test remote_nodes/test_mirror/
3. Watch the logs in the server as it's being called by the test and as they call each other!

Automated testing does a lot more checks/assertions both in the calling and receiving servers.

## Automated testing in action
With the two servers running, just run the test and watch as the calls happen

![automated test](https://user-images.githubusercontent.com/21351900/203461358-a92d5871-54a6-4718-bcc6-d387d75b850a.gif)

![image](https://user-images.githubusercontent.com/21351900/203461378-447bd854-012e-4fba-a025-dac751c5a809.png)


## The full manual testing flow
If you want to but really 🤷‍♂️

1. Follow an author: `POST http://127.0.0.1:8000/authors/<8080 author>/followers`
2. See incoming requests: `POST http://127.0.0.1:8080/follows/incoming/`
3. Accept a follow request: `PUT http://127.0.0.1:8000/authors/<8080 author>/followers/<8000 author>` with payload `{"hasAccepted": true}`
5. Check if following: `GET http://127.0.0.1:8000/authors/<8080 author>/followers/<8000 author>` or iterate through `GET http://127.0.0.1:8000/authors/<8080 author>/followers/`
6. Unfollow someone: `GET http://127.0.0.1:8000/authors/<8080 author>/followers/<8000 author>`
